### ISSUE TITLE
Structured grid dimensions should be immutable

### ISSUE THREAD

**Issue by akaszynski** [pyvista/pyvista#4135]
Attempting to modify the dimensions of a `StructuredGrid` corrupts it:

```py
import pyvista
import numpy as np
xrng = np.arange(-10, 10, 1, dtype=np.float32)
yrng = np.arange(-10, 10, 2, dtype=np.float32)
zrng = np.arange(-10, 10, 5, dtype=np.float32)
x, y, z = np.meshgrid(xrng, yrng, zrng)
grid = pyvista.StructuredGrid(x, y, z)
grid.dimensions = (3, 4, 5)
grid.plot()
```
![tmp3](https://user-images.githubusercontent.com/11981631/225195301-fc9624c4-5684-4b3f-9e6b-06a175910140.png)

Whereas for a `UniformGrid`, it does not:

```py
ugrid = pyvista.UniformGrid(dimensions=(4, 5, 6))
ugrid.dimensions = (3, 3, 3)
ugrid.plot(show_edges=True)
```

![tmp2](https://user-images.githubusercontent.com/11981631/225195235-940e75f0-cb81-4a82-9aa7-0a524ef4662d.png)

**Comment by banesullivan** [comment]
The dimensions should be mutable! These are a property that can change, and there are lots of examples out there where the dimensions get re-set. The problem in your example is that your new dimensions are wrong!

Further, the example with a uniform grid is completely different and not a fair comparison. The uniform grid is defined by its dimensions. Changing the dimensions produces an entirely different mesh. Whereas changing the dimensions of the structured grid means something totally different: how to interpret the structured points.

If this property were made immutable, it will break tests, examples, and lots of downstream code.

One such example is later on in: https://docs.pyvista.org/examples/00-load/create-structured-surface.html

```py
mesh = pv.StructuredGrid()
# Set the coordinates from the numpy array
mesh.points = points
# set the dimensions
mesh.dimensions = [29, 32, 1]
```

And at least one other:

https://docs.pyvista.org/examples/00-load/create-surface-draped.html#sphx-glr-examples-00-load-create-surface-draped-py

Further, I often am given structured points from external software where the dimensions are unknown, and I need to be able to toy with the dimensions (basically guess) until the mesh looks right.

Additionally, there are definitely scenarios where a point set may be valid under multiple dimensions -- I don't have a good example off hand, but I'm pretty sure I've seen this before

**Comment by adeak** [comment]
@banesullivan do you have a recommendation for how we can do something about this? We could just test the consistency of values in the `points` setter XOR the `dimensions` setter, but then we'd force users to always set one before the other. But if both are set independently then we can't really put verification logic anywhere unless we define a consistency verification method on datasets which get called by plotting routines. Any other options?

**Comment by banesullivan** [comment]
`nx*ny*nz == n_points` would be the best check IMO

**Comment by adeak** [comment]
I kind of figured that :P The question is where and when to check.

**Comment by banesullivan** [comment]
However, you're right that this would require users to set the points before the dimensions. IMO, this whole thing should be left as-os

**Comment by banesullivan** [comment]
I just flat out don't think we should add any additional constraints here. VTK doesn't and we should honor that.

... I know I've said the complete opposite in other scenarios like setting plotting parameters that don't make sense 🙃

**Comment by banesullivan** [comment]
IMO, the dataset API should be as close to VTK as possible with minimal extra checks. The plotting API is where I think we have more freedom to add these sorts of sanity checks. 

VTK power users (really speaking for myself here) wouldn't be thrilled if the dimensions were constrained

**Comment by adeak** [comment]
> VTK power users (really speaking for myself here) wouldn't be thrilled if the dimensions were constrained

What I have in mind is constraining them right before you shoot yourself in the foot with an inconsistent grid. It wouldn't make sense to plot an inconsistent grid, for instance.

**Comment by banesullivan** [comment]
I'd be open to have some "pre-use hook method" on all datasets that does these sorts of checks. This could run before plotting (and maybe even filters). But overall, I'm mostly of the opinion that this should be left alone.... maybe I want crazy spiky datasets!

**Comment by banesullivan** [comment]
> but this introduces more moving parts

Exactly... IMO, this is okay for the plotting API but I think we should avoid this in the core API

**Comment by adeak** [comment]
The only reason I wondered about pushing this to the grids themselves is to reduce the overhead during plotting that comes from checking for very rare, mostly silly cases. But you're right that it's better to keep this tucked away wherever it's needed, and in any case the consistency check is very cheap (just a few attribute lookups and integer arithmetic operations and comparisons).
